pallets/protocol-fee/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+
+type BalanceOf<T> =
+    <<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::Balance;
+type NegativeImbalanceOf<T> =
+    <<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::NegativeImbalance;
+/// Either the computed fee or an error.
+pub type ComputeFeeResult<T> = sp_std::result::Result<BalanceOf<T>, DispatchError>;
+/// Either an imbalance or an error.
+type WithdrawFeeResult<T> = sp_std::result::Result<NegativeImbalanceOf<T>, DispatchError>;
+
+pub trait Trait: frame_system::Trait {
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+    /// The currency type in which fees will be paid.
+    type Currency: Currency<Self::AccountId> + Send + Sync + IdentityCurrency<Self::AccountId>;
+    /// Handler for the unbalanced reduction when taking protocol fees.
+    type OnProtocolFeePayment: OnUnbalanced<NegativeImbalanceOf<Self>>;
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// Insufficient identity balance to pay the fee.
+        InsufficientIdentityBalance,
+        /// Insufficient account balance to pay the fee.
+        InsufficientAccountBalance,
+        /// Account ID decoding failed.
+        AccountIdDecode,
+        /// Division in `compute_fee` failed.
+        ComputeFee,
+    }
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as ProtocolFee {
+        /// The mapping of operation names to the base fees of those operations.
+        pub BaseFees get(base_fees) config(): map ProtocolOp => BalanceOf<T>;
+        /// The fee coefficient as a positive rational (numerator, denominator).
+        pub Coefficient get(coefficient) config() build(|config: &GenesisConfig<T>| {
+            if config.coefficient.1 == 0 {
+                PosRatio(1, 1)
+            } else {
+                config.coefficient
+            }
+        }): PosRatio;
+    }
+}
+
+decl_event! {
+    pub enum Event<T> where Balance = BalanceOf<T> {
+        /// The protocol fee of an operation.
+        Fee(Balance),
+        /// The fee coefficient.
+        Coefficient(PosRatio),
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Changes the fee coefficient for the root origin.
+        pub fn change_coefficient(origin, coefficient: PosRatio) -> DispatchResult {
+            ensure_root(origin)?;
+            <Coefficient>::put(coefficient);
+            Ok(())
+        }
+
+        /// Changes the a base fee for the root origin.
+        pub fn change_base_fee(origin, op: ProtocolOp, base_fee: BalanceOf<T>) ->
+            DispatchResult
+        {
+            ensure_root(origin)?;
+            <BaseFees<T>>::insert(op, base_fee);
+            Ok(())
+        }
+
+        /// Emits an event with the fee of the operation.
+        pub fn get_fee(_origin, op: ProtocolOp) -> DispatchResult {
```

Comment:
I guess we will need to wrap in an RPC function for easier use by the dApp team.

Reply:
tbh, it's so simple to calculate the fees that even on the dApp it might be easier to read raw fee and coefficient and from storage and just multiplying them.